lib/Sharing/DeckShareProvider.php

```diff
@@ -267,7 +267,13 @@ private function createShareObject(array $data): IShare {
 		return $share;
 	}
 
-	private function applyBoardPermission($share, $permissions, $userId) {
+	private function applyBoardPermission($share, $permissions, $userId, ?array $shareToPermissionMap = null): void {
+		$cardId = (int)$share->getSharedWith();
+		if ($shareToPermissionMap !== null && isset($shareToPermissionMap[$cardId])) {
+			$share->setPermissions($this->permissionService->boardToFilePermission($shareToPermissionMap[$cardId]));
+			return;
+		}
+
 		try {
 			$this->permissionService->checkPermission($this->cardMapper, $share->getSharedWith(), Acl::PERMISSION_EDIT, $userId);
```

Comment:
This is the previously expensive part as it needs to:
- find the board id for the card which we already have in the result of the query that fetches all shares
- fetches the permissions for each board (N+1) which can be optimized by only fetching once we have all board ids
- Then a map from card id to the related board permissions is built and passed along to the logic above to actually build the file permissions